docs/research/blake-256-hash-function.md

```diff
@@ -1,13 +1,302 @@
+round **r**, **M<sub>σ<sub>r</sub></sub>** are message blocks and **C<sub>σ<sub>r</sub></sub>**
+are round-dependent constants (see appendix for more information on permutations and constants). The **G<sub>i</sub>** updates 4, 32-bit state words (a, b, c, d) twice as follows:
+
+
+>a ← a + b + (m<sub>σ<sub>r</sub></sub>(2i) ⊕ c<sub>σ<sub>r</sub></sub>(2i+1))
+
+>d ← (d ⊕ a) >>> 16
+
+>c ← c + d
+
+>b ← (b ⊕ c) >>> 12
+
+>a ← a + b + (m<sub>σ<sub>r</sub></sub>(2i+1) ⊕ c<sub>σ<sub>r</sub></sub>(2i))
+
+>d ← (d ⊕ a) >>> 8
+
+>c ← c + d
+
+>b ← (b ⊕ c) >>> 7
+
+**<em>Figure 4: G Function, based on ChaCha [^1]</em>**
+
+|Symbol|Meaning|
+|------|-------|
+|←|Variable assignment|
+|+|Addition Modulo 2<sup>32</sup>|
+|⊕|Boolean exclusive OR (XOR)|
+|>>>|k K-bit rotation to the right|
+|<<<|k K-bit rotation to the left
+
+**<em>Table 2: Operations symbols used to describe the **G** function</em>**
+
+The **G** function is inspired from the “quarter-round” function of the stream cipher ChaCha, 
+with the difference being the addition of the message words, and the reversing of the rotation directions. ChaCha transforms (a, b, c, d) as follows:
+
+>a ← a + b
+
+>d ← (d ⊕ a) <<< 16
+
+>c ← c + d
+
+>b ← (b ⊕ c) <<< 12
+
+>a ← a + b
+
+>d ← (d ⊕ a) <<< 8
+
+>c ← c + d
+
+>b ← (b ⊕ c) <<< 7
+
+**<em>Figure 5: ChaCha Function [^1]</em>**
+
+### **Security Evaluation Criteria**
+
+The security of a cryptographic hash function is determined by the number of queries or guesses required to solve the following problems:
+
+1) Pre-image resistance (one-way): Given the **_y_** as an output of the hash function, it is computationally infeasible to find message **_x_** such that **_h(x) = y_**. [^4]
+
+![Preimage resistance](/img/Preimage resistance.svg)
+
+**<em>Figure 6: Pre-image resistance </em>**
+
+2) Second pre-image resistance (weak collision resistant): Given **_x_**, it is computationally infeasible to find a second pre-image  **_x' ≠ x_** such that **_h(x) = h(x')_** [^4]
+
+![Second pre-image resistance](/img/Second preimage resistance.svg)
+
+**<em>Figure 7: Second pre-image resistance </em>**
+
+3) Collision resistance (strong collision-resistant): It is computationally infeasible to find any two distinct inputs **_x, x'_** that hash to the same output such that **_h(x) = h(x')_** [^4]
+
+![Collision resistance](/img/Collision resistance.svg)
+
+**<em>Figure 8: Collision resistance </em>**
+
+All three properties contain the phrase _computationally infeasible_, meaning it would take the fastest computer a long time to solve the problem, e.g., 100 to 10000 years, making it impossible in practice. In addition to the above three properties, resistance to Length-extension attacks is an additional requirement which outlined in NIST's evaluation criteria for its SHA-3 candidates hash function [^15]:
```

Comment:
This is a personal thing, but I don't agree with calling a 100 year run time "computationally infeasible". Even with a pessimistic view of Moore's law, something which takes 100 years to execute today could potentially be executed in a matter of hours in a decade or two.

To me, "computationally infeasible" is in the order of billions of years.